mlir/include/mlir/Dialect/Math/Transforms/Passes.td

```diff
@@ -36,4 +36,51 @@ def MathLegalizeToF32 : Pass<"math-legalize-to-f32"> {
   let dependentDialects = ["math::MathDialect", "arith::ArithDialect"];
 }
 
+def MathCanonicalizeF32Promotion : Pass<"math-canonicalize-f32-promotion"> {
+  let summary = "Eliminate redundant truncf/extf pairs";
+  let description = [{
+    `legalize-to-f32` pass does f32 promotion for every op belonging to the
+    illegal op list. Once there are some consecutive illegal ops, `legalize-to-f32`
+    will insert redundant `arith.truncf` and `arith.extf` pairs between the illegal
+    ops.
+    
+    This pass is to eliminate the redundant truncf/extf pairs to improve
+    performance.
+
+    However, this pass may introduce numerical difference as the `f32->bf16` rounding
```

Comment:
Would these be valid patterns as canonicalization when fast-math is enabled?

Reply:
This pass is independent from fast-math. If the fast-math is enabled, and we have such redundant truncf/extf pairs, they will still be removed from the IR.

Reply:
I think you missed my point: I was asking about THE canonicalization patterns, not a separate pass.

Reply:
Given the context of how that ultimately got resolved, what I'd want to see is updating the math legalizer to strip out trunc/ext pairs *that it creates* at time of creation.

Reply:
> I don't think they're valid under fastmath.

Can you elaborate? I don't follow. 
Assuming you meant to send a link to https://github.com/llvm/llvm-project/pull/88486#discussion_r1586918033 instead, there you wrote:

> This rewrite has caused per-element result errors around 1e-2 (if I remember right 16.25 vs 16.3125 or the like)

But numerical differences is actually in scope for fast-math...

Reply:
No, I did mean to send a link to the commit - the discussion moved there.

What I mean to say is that
```mlir
%y = arith.truncf %arg0 : float to half
%z =arith.extf %y : half to float
return %z : float
```

can't be simplified, even under fastmath, because that is an explicit, user-specified desire to lose precision.

However, rewrites (like `arith-emulate-unsupported-floats`, which probably should also get this intermediate preservation treatment) which would introduce such a colliding truncf/extf pair are allowed to not do that and keep a higher intermediate precision. This is allowed without fastmath.

These are, to my current understanding, the C, and thus the LLVM, semantics.

Reply:
If we didn't have this wider ecosystem that knows what exactly the fastmath flags mean, I'd buy that you could do this rewrite everywhere under `contract` ... but I'm pretty sure that flag doesn't allow eliminating arbitrary trunc/ext pairs from input

Reply:
> can't be simplified to return %arg0, even under fastmath, because that is an explicit, user-specified desire to lose precision.

Sorry but I don't quite see why fast-math does not allow this in LLVM IR?
I don't really buy the "user-specified desired" because patterns of IR are happening in LLVM after quite a churn of optimization (possibly only through inlining, etc.). 
If the user *really* want something, they shouldn't decorate these with fast math!

Reply:
I think the main thing with LLVM IR is hat `fptrunc` and `fpext` can't carry fastmath flags, last I checked.

Now, if they could, I'd absolutely permit
```mlir
func.func @f(%arg0: f32) -> f32 {
    %0 = arith.truncf contract %arg0 : f32 to f16
    %1 = arith.extf contract %0 : f16 to f32
    return %1 : f32
}
```
to fold to
```mlir
func.func @f(%arg0: f32) -> f32 {
    return %arg0 : f32
}
```

So, from an MLIR perspective, if we give extend and truncate the ability to fastmath (if they don't already have it) it'd make a lot of sense to
1. Add that folding and
2. Change the various unsupported float emulation/legalization passes to (optionally - or maybe by default) stick a `contract` on their truncates and extensions